### PR TITLE
Change type column of object_relations_* tables to enum

### LINKED ISSUE #10662: [Improvement]: Change `type` column of `object_relations_*` tables to `enum`
### Improvement description

for smaller footprint ...

### PR DESCRIPTION
**Pull request by BabovicT**
Resolves #10662

### CODE REVIEW

File: bundles/CoreBundle/Migrations/Version20211103055110.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace App\Migrations\SixNine;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+class Version20211103055110 extends AbstractMigration
```
Comment (matthiashamacher) [inline bundles/CoreBundle/Migrations/Version20211103055110.php:26]:
```suggestion
/**
 * @internal
 */
class Version20211103055110 extends AbstractMigration
```

File: bundles/CoreBundle/Migrations/Version20211103055110.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace App\Migrations\SixNine;
```
Comment (matthiashamacher) [inline bundles/CoreBundle/Migrations/Version20211103055110.php]: Please correct the namespace

File: bundles/CoreBundle/Migrations/Version20211103055110.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
+<?php
+
```
Comment (matthiashamacher) [inline bundles/CoreBundle/Migrations/Version20211103055110.php:2]: Please add the pimcore copyright and license info like here: https://github.com/pimcore/pimcore/blob/f1488674f12bb9f79f157c167337bff4ae2240fd/bundles/CoreBundle/Migrations/Version20201014101437.php#L3-L14

File: bundles/CoreBundle/Migrations/Version20211103055110.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
+<?php
```
Comment (matthiashamacher) [inline bundles/CoreBundle/Migrations/Version20211103055110.php:1]: Please add strict type declaration
```suggestion
<?php

declare(strict_types=1);
```

File: bundles/CoreBundle/Migrations/Version20211103055110.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+namespace App\Migrations\SixNine;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+class Version20211103055110 extends AbstractMigration
+{
+    /**
+     * @param Schema $schema
+     */
```
Comment (matthiashamacher) [inline bundles/CoreBundle/Migrations/Version20211103055110.php]:
```suggestion
```

File: bundles/CoreBundle/Migrations/Version20211103055110.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+                    "ALTER TABLE $objectDatastoreTableRelation CHANGE COLUMN 
+                        `type` `type` ENUM('object', 'asset', 'document') NOT NULL DEFAULT 'object' ;"
+                );
+            }
+        }
+    }
+
+    /**
+     * @param Schema $schema
+     */
```
Comment (matthiashamacher) [inline bundles/CoreBundle/Migrations/Version20211103055110.php]:
```suggestion

```

**Review by matthiashamacher — CHANGES_REQUESTED** [review]

File: bundles/CoreBundle/Migrations/Version20211103055110.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+ * Full copyright and license information is available in
+ * LICENSE.md which is distributed with this source code.
+ *
+ *  @copyright  Copyright (c) Pimcore GmbH (http://www.pimcore.org)
+ *  @license    http://www.pimcore.org/license     GPLv3 and PCL
+ */
+
+/**
+ * @internal
+ */
```
Comment (matthiashamacher) [inline bundles/CoreBundle/Migrations/Version20211103055110.php]: This needs to go above the class not namespace

```suggestion
namespace Pimcore\Bundle\CoreBundle\Migrations;

use Doctrine\DBAL\Schema\Schema;
use Doctrine\Migrations\AbstractMigration;

/**
 * @internal
 */
 class Version20211103055110 extends AbstractMigration
```

**Review by matthiashamacher — CHANGES_REQUESTED** [review]

File: bundles/CoreBundle/Migrations/Version20211103055110.php (hunk 7 above)
Comment (BabovicT) [reply bundles/CoreBundle/Migrations/Version20211103055110.php]: Sry :(

**Review by matthiashamacher — APPROVED** [review]
LGTM

File: bundles/CoreBundle/Migrations/Version20211103055110.php (hunk 7 above)
Comment (matthiashamacher) [reply bundles/CoreBundle/Migrations/Version20211103055110.php]: No worries :)

File: models/DataObject/ClassDefinition/Dao.php [hunk 8]
```diff
@@ -146,7 +146,7 @@ public function update()
               `id` BIGINT(20) NOT NULL PRIMARY KEY  AUTO_INCREMENT,
               `src_id` int(11) UNSIGNED NOT NULL DEFAULT '0',
               `dest_id` int(11) UNSIGNED NOT NULL DEFAULT '0',
-              `type` varchar(50) NOT NULL DEFAULT '',
+              `type` enum('object', 'asset','document') NOT NULL DEFAULT 'object',
```
Comment (brusch) [inline models/DataObject/ClassDefinition/Dao.php]:
```suggestion
              `type` enum('object', 'asset','document') NULL DEFAULT NULL,
```
I'd let `NULL` be the default value, as `object` could lead to unexpected behavior. 
We'd then also need to change the migration script: 
- let `type` be `NULL`
- change all `''` (empy values)  to `NULL`
- update the column definition to enum

**Review by brusch — CHANGES_REQUESTED** [review]
See my inline comment 😊
Would be nice if you could update the PR accordingly. Thanks in advance!

File: models/DataObject/ClassDefinition/Dao.php (hunk 8 above)
Comment (BabovicT) [reply models/DataObject/ClassDefinition/Dao.php]: @brusch like that?

Comment (brusch) [reply models/DataObject/ClassDefinition/Dao.php]: Exactly - thanks a  lot 👍

**Pull request merged**